azure_provisioning_e2e/tests/test_sync_certificate_enrollments.py

```diff
@@ -0,0 +1,266 @@
+    call_device_cert_creation_from_pipeline,
+    delete_directories_certs_created_from_pipeline,
+)
+
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+intermediate_common_name = "e2edpswingardium"
+intermediate_password = "leviosa"
+device_common_name = "e2edpsexpecto"
+device_password = "patronum"
+
+service_client = ProvisioningServiceClient.create_from_connection_string(
+    os.getenv("PROVISIONING_SERVICE_CONNECTION_STRING")
+)
+device_registry_helper = Helper(os.getenv("IOTHUB_CONNECTION_STRING"))
+PROVISIONING_HOST = os.getenv("PROVISIONING_DEVICE_ENDPOINT")
+ID_SCOPE = os.getenv("PROVISIONING_DEVICE_IDSCOPE")
+
+
+@pytest.fixture(scope="module")
+def before_all_tests(request):
+    logging.info("set up certificates before cert related tests")
+    call_intermediate_cert_creation_from_pipeline(
+        common_name=intermediate_common_name,
+        ca_password=os.getenv("PROVISIONING_ROOT_PASSWORD"),
+        intermediate_password=intermediate_password,
+    )
+    call_device_cert_creation_from_pipeline(
+        common_name=device_common_name,
+        intermediate_password=intermediate_password,
+        device_password=device_password,
+        device_count=8,
+    )
+
+    def after_module():
+        logging.info("tear down certificates after cert related tests")
+        delete_directories_certs_created_from_pipeline()
+
+    request.addfinalizer(after_module)
+
+
+@pytest.mark.it(
+    "A device gets provisioned to the linked IoTHub with the user supplied device_id different from the registration_id of the individual enrollment that has been created with a selfsigned X509 authentication"
+)
+def test_device_register_with_device_id_for_a_x509_individual_enrollment(before_all_tests):
+    device_id = "e2edpsflyingfeather"
+    device_index = 1
+
+    try:
+        individual_enrollment_record = create_individual_enrollment_with_x509_client_certs(
+            device_index=device_index, device_id=device_id
+        )
+        registration_id = individual_enrollment_record.registration_id
+
+        device_cert_file = "demoCA/newcerts/device_cert" + str(device_index) + ".pem"
+        device_key_file = "demoCA/private/device_key" + str(device_index) + ".pem"
+        result_from_register(registration_id, device_cert_file, device_key_file)
+
+        assert_device_provisioned(device_id)
+    finally:
+        service_client.delete_individual_enrollment_by_param(registration_id)
+
+
+@pytest.mark.it(
+    "A device gets provisioned to the linked IoTHub with device_id equal to the registration_id of the individual enrollment that has been created with a selfsigned X509 authentication"
+)
+def test_device_register_with_no_device_id_for_a_x509_individual_enrollment(before_all_tests):
+
+    device_index = 2
+
+    try:
+        individual_enrollment_record = create_individual_enrollment_with_x509_client_certs(
+            device_index=device_index
+        )
+        registration_id = individual_enrollment_record.registration_id
```

Comment:
Conversely, here you probably should assert that the registration id is the same as the device id

Reply:
going to the assert method (line 100 and line 231) we are actually doing 
`assert device.device_id == device_id` but in this case the registration_id is passed , so the actual assert becomes `assert device.device_id == registration_id`

---
In reply to: [314917762](https://github.com/Azure/azure-iot-sdk-python-preview/pull/167#discussion_r314917762) [](ancestors = 314917762)